### PR TITLE
Support Storage Capacity Tracking

### LINKED ISSUE #193: Add mode for Storage Capacity Tracking
## What

[Storage Capacity Tracking](https://github.com/kubernetes/enhancements/issues/1472) is a new alpha feature of Kubernetes 1.19.

The feature enables us to track volume group capacities and filter nodes not having sufficient storage capacity from pod scheduling without `topolvm-scheduler`.

A lot of users will gain benefits from this because:
- Users of managed Kubernetes are often not able to edit `kube-scheduler` configurations.
- Changing `kube-scheduler` configuration is troublesome and difficult.

So, in addition to using the scheduler extender, add a new mode to use storage capacity tracking.

closes #235 

## How

Describe how to address the issue.

## Checklist

- [ ] Finish implementation of the issue
- [ ] Test all functions
- [ ] Have enough logs to trace activities
- [ ] Notify developers of necessary actions

### PR DESCRIPTION
**Pull request by bells17**
resolve #193

### CODE REVIEW

File: e2e/scheduler_test.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ var capacityPodTemplateYAML string
 func testScheduler() {
 	var cc CleanupContext
 	BeforeEach(func() {
+		skipIfStorageCapacity()
```
Comment (toshipp) [inline e2e/scheduler_test.go:20]: I think we have to confirm pods are scheduled using capacity tracking in an end-to-end manner.
Or just adding a unit test for GetCapacity is sufficient.
I want your thought.

File: deploy/manifests/overlays/storage-capacity/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+  - ../../base
+patchesStrategicMerge:
+  - controller.yaml
+images:
+  - name: quay.io/topolvm/topolvm-with-sidecar
+    newTag: 0.8.2
```
[the remark this answers is not in the document]
Comment (bells17) [reply deploy/manifests/overlays/storage-capacity/kustomization.yaml]: Sorry, I fixed it.
https://github.com/topolvm/topolvm/pull/315/commits/9ff43d6b5d4f0c50a0213cc448fff6f22b2cc1cd

File: e2e/Makefile [hunk 3]
```diff
@@ -32,8 +32,21 @@ ifeq ($(TEST_SCHEDULER_MANIFEST),deployment)
 KUSTOMIZE_DIR=manifests/overlays/deployment-scheduler
 endif
 
+ifeq ($(KUBERNETES_VERSION),1.21.0)
+KUSTOMIZE_DIR=manifests/overlays/storage-capacity
+endif
```
[the remark this answers is not in the document]
Comment (bells17) [reply e2e/Makefile:40]: fixed it.
https://github.com/topolvm/topolvm/pull/315/commits/c4223dc6705b3d398f3edc410bc3b241ce7f08ee

File: e2e/hook_test.go [hunk 4]
```diff
@@ -35,6 +35,7 @@ func hasTopoLVMFinalizer(pvc *corev1.PersistentVolumeClaim) bool {
 func testHook() {
 	var cc CleanupContext
 	BeforeEach(func() {
+		skipIfStorageCapacity()
```
[the remark this answers is not in the document]
Comment (bells17) [reply e2e/hook_test.go]: fixed it.
https://github.com/topolvm/topolvm/pull/315/commits/512c9720de7a3b8a1fe259ab5e3853b3437bc7e1

File: e2e/manifests/overlays/daemonset-lvmd-storage-capacity/mutatingwebhooks.yaml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+$patch: delete
```
[the remark this answers is not in the document]
Comment (bells17) [reply e2e/manifests/overlays/daemonset-lvmd-storage-capacity/mutatingwebhooks.yaml]: fixed it.
https://github.com/topolvm/topolvm/pull/315/commits/22cead8102963746379f5dbd79fbb41e0de03064

File: e2e/scheduler_test.go (hunk 1 above)
Comment (bells17) [reply e2e/scheduler_test.go]: > I think we have to confirm pods are scheduled using capacity tracking in an end-to-end manner.
> Or just adding a unit test for GetCapacity is sufficient because scheduling must be tested by kubernetes.
> May I ask you your thought?

Um...I think storage capacity scheduling tests are written sufficiently by Kubernetes like below:
- https://github.com/kubernetes/kubernetes/blob/v1.21.1/pkg/controller/volume/scheduling/scheduler_binder_test.go#L2202-L2354
- https://github.com/kubernetes/kubernetes/blob/v1.21.1/test/e2e/storage/testsuites/capacity.go

So I think it's enough to add tests for setting the CSIStorageCapacity resource by the csi-provisioner and topolvm GetCapacity.
What do you think?

Comment (bells17) [reply e2e/scheduler_test.go]: I noticed that can't write test cases for storage capacity tracking because k8s.io/api package version is v0.20.2 currently...
It is necessary to upgrade k8s.io/api and controller-runtime to use CSIStorageCapacity resource in e2e testing.

File: driver/controller.go [hunk 6]
```diff
@@ -96,7 +96,7 @@ func (s controllerService) CreateVolume(ctx context.Context, req *csi.CreateVolu
 			return nil, status.Error(codes.Internal, "can not find any node")
 		}
 		if capacity < (requestGb << 30) {
-			return nil, status.Errorf(codes.Internal, "can not find enough volume space %d", capacity)
+			return nil, status.Errorf(codes.ResourceExhausted, "can not find enough volume space %d", capacity)
```
Comment (bells17) [inline driver/controller.go:99]: https://kubernetes-csi.github.io/docs/storage-capacity-tracking.html
> - return the ResourceExhausted gRPC status code in CreateVolume if capacity is exhausted
> - use external-provisioner >= 1.6.0 because older releases did not properly support rescheduling after a ResourceExhausted error

File: driver/controller.go [hunk 7]
```diff
@@ -242,12 +242,15 @@ func (s controllerService) GetCapacity(ctx context.Context, req *csi.GetCapacity
 		var err error
 		capacity, err = s.nodeService.GetTotalCapacity(ctx, deviceClass)
 		if err != nil {
-			return nil, status.Error(codes.Internal, err.Error())
+			ctrlLogger.Error(err, "get total capacity failed")
+			return &csi.GetCapacityResponse{AvailableCapacity: 0}, nil
```
Comment (bells17) [inline driver/controller.go:246]: https://kubernetes.slack.com/archives/C09QZFCE5/p1622273728215500
> My expectation was that a CSI driver returns "no capacity" = 0 when it cannot create volumes with the requested parameters.
> As you said, there is no specific error code for this scenario whereas 0 is a valid and correct answer.

File: csi-sidecars.mk [hunk 8]
```diff
@@ -1,5 +1,5 @@
 # CSI sidecar versions
-EXTERNAL_PROVISIONER_VERSION = 2.2.0
+EXTERNAL_PROVISIONER_VERSION = 495c3270db82b0bc282d15aab0a1c1c317596e9a
```
Comment (bells17) [inline csi-sidecars.mk:2]: https://github.com/kubernetes-csi/external-provisioner/pull/635

File: e2e/scheduler_test.go (hunk 1 above)
Comment (toshipp) [reply e2e/scheduler_test.go]: I agreed to add a test for the CSIStorageCapacity resource.

File: driver/controller.go (hunk 7 above)
Comment (toshipp) [reply driver/controller.go:246]: I understand it is expected not an error. Nice clarification.

File: deploy/README.md [hunk 9]
```diff
@@ -253,7 +255,7 @@ spec:
                     operator: Exists
 ```
 
-### Running topolvm-scheduler using Deployment and Service
+#### Running topolvm-scheduler using Deployment and Service
```
Comment (toshipp) [inline deploy/README.md:258]: `OPTIONAL: tune the node scoring` section is level2 heading, but it is only related to topolvm-scheduler.
Would you make it smaller?

File: deploy/README.md [hunk 10]
```diff
@@ -300,6 +302,28 @@ extenders:
     ignoredByScheduler: true
 ```
 
+### Storage Capacity Tracking
+
+topolvm supports [Storage Capacity Tracking](https://kubernetes.io/docs/concepts/storage/storage-capacity/).
+You can enable Storage Capacity Tracking mode instead of using topolvm-scheduler.
+You need to use Kubernetes Cluster v1.21 ~ when using Storage Capacity Tracking with topolvm.
```
Comment (toshipp) [inline deploy/README.md:309]:
```suggestion
You need to use Kubernetes Cluster v1.21 or later when using Storage Capacity Tracking with topolvm.
```

File: driver/controller.go (hunk 7 above)
Comment (toshipp) [reply driver/controller.go:246]: GetTotalCapacity returns a real error if getNodes failed.
https://github.com/topolvm/topolvm/blob/c974ca0f99e497a83b450ccdae369d0262962f48/driver/k8s/node_service.go#L80
And also, if deviceClass annotation is missing, this code returns 0, so this change is not needed, IMO.
https://github.com/topolvm/topolvm/blob/c974ca0f99e497a83b450ccdae369d0262962f48/driver/k8s/node_service.go#L85-L90

File: driver/controller.go [hunk 11]
```diff
@@ -257,7 +260,8 @@ func (s controllerService) GetCapacity(ctx context.Context, req *csi.GetCapacity
 			return &csi.GetCapacityResponse{AvailableCapacity: 0}, nil
 		case nil:
 		default:
-			return nil, status.Error(codes.Internal, err.Error())
+			ctrlLogger.Error(err, "get capacity failed")
+			return &csi.GetCapacityResponse{AvailableCapacity: 0}, nil
```
Comment (toshipp) [inline driver/controller.go:264]: GetCapacityByTopologyLabel calls getNodes and the method would return some real error.
Would you revise the code to handle errors more precisely?

File: e2e/Makefile [hunk 12]
```diff
@@ -192,14 +219,28 @@ daemonset-lvmd/delete-minikube:
 # Set scheduler configs manually because minikube can't edit scheduler configs.
 .PHONY: daemonset-lvmd/update-minikube-setting
 daemonset-lvmd/update-minikube-setting: daemonset-lvmd/delete-minikube daemonset-lvmd/launch-minikube
+ifeq ($(STORAGE_CAPACITY),false)
```
Comment (toshipp) [inline e2e/Makefile:222]: It would be better to use `ifneq` or comparing to `true`, IMO.

**Review by toshipp — CHANGES_REQUESTED** [review]

File: deploy/README.md (hunk 9 above)
Comment (bells17) [reply deploy/README.md:258]: fixed it.
https://github.com/topolvm/topolvm/blob/17c094b979ecf53030ff1850a8b8f923e7264a9e/deploy/README.md

File: e2e/Makefile (hunk 12 above)
Comment (bells17) [reply e2e/Makefile]: improved it.
https://github.com/topolvm/topolvm/pull/315/commits/eb48cda127de7da48b550aa58f7236ab03777a6d

File: driver/k8s/node_service.go [hunk 13]
```diff
@@ -68,7 +69,11 @@ func (s NodeService) GetCapacityByTopologyLabel(ctx context.Context, topology, d
 			if v != topology {
 				continue
 			}
-			return s.extractCapacityFromAnnotation(&node, dc)
+			c, err := s.extractCapacityFromAnnotation(&node, dc)
+			if err != nil {
+				return 0, ErrDeviceClassNotFound
```
Comment (toshipp) [inline driver/k8s/node_service.go:74]: In this function, there is no device class check, and it is actually in extractCapacityFromAnnotation.
Would you move this code to extractCapacityFromAnnotation?

**Review by toshipp — CHANGES_REQUESTED** [review]

File: e2e/capacity_test.go [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+				return fmt.Errorf("%v: stdout=%s, stderr=%s", err, result, stderr)
+			}
+			if !strings.Contains(string(podInfo), "did not have enough free storage") {
+				return errors.New("scheduling failed message is not found")
+			}
+
+			return nil
+		}).Should(Succeed())
+
+		if !isDaemonsetLvmdEnvSet() {
```
Comment (toshipp) [inline e2e/capacity_test.go:144]: Would you explain why this check is needed? This is not related to daetmonset lvmd, IMO.

File: e2e/Makefile [hunk 15]
```diff
@@ -190,19 +208,33 @@ daemonset-lvmd/launch-minikube:
...
+ifneq ($(STORAGE_CAPACITY),true)
 	$(SUDO) mkdir -p /var/lib/scheduler
 	$(SUDO) sh -c 'sed -e "s|@DEPLOYMENT_SCHEDULER_HOST@|127.0.0.1|" $(SCHEDULER_CONFIG) > /var/lib/scheduler/scheduler-config.yaml'
 	$(SUDO) sh -c 'sed -e "s|@KUBERNETES_VERSION@|$(KUBERNETES_VERSION)|" kube-scheduler.yaml > /etc/kubernetes/manifests/kube-scheduler.yaml'
+else
+endif
 
 .PHONY: daemonset-lvmd/test
 daemonset-lvmd/test: topolvm.img $(SERVER_CERT_FILES)
+ifneq ($(STORAGE_CAPACITY),true)
```
Comment (llamerada-jp) [inline e2e/Makefile:225]: It would be simpler to write it the same as KUSTOMIZE_DIR.

File: e2e/capacity_test.go [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+			var nodelist corev1.NodeList
+			result, _, err := kubectl("get", "node", "-o=json")
+			Expect(err).ShouldNot(HaveOccurred())
+
+			err = json.Unmarshal(result, &nodelist)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			if len(nodelist.Items) != 1 {
+				Expect(errors.New("count of node is invalid")).ShouldNot(HaveOccurred())
+			}
```
Comment (llamerada-jp) [inline e2e/capacity_test.go:46]:
```suggestion
                      Expect(nodelist.Items).Should(HaveLen(1), "count of node is invalid")
```

File: e2e/capacity_test.go [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+		Expect(err).ShouldNot(HaveOccurred())
+
+		var node corev1.Node
+		err = json.Unmarshal(result, &node)
+		Expect(err).ShouldNot(HaveOccurred())
+		size := func() string {
+			sizeStr, exists := node.Annotations["capacity.topolvm.cybozu.com/00default"]
+			if !exists {
+				Expect(errors.New("size is not found")).ShouldNot(HaveOccurred())
+			}
```
Comment (llamerada-jp) [inline e2e/capacity_test.go:62]:
```suggestion
			Expect(exists).Should(BeTrue(), "size is not found")
```

**Review by llamerada-jp — CHANGES_REQUESTED** [review]

File: e2e/Makefile (hunk 15 above)
Comment (bells17) [reply e2e/Makefile]: I fixed it.
https://github.com/topolvm/topolvm/pull/315/commits/b3e0d4db8f07fe24445ebcbbce16b24b698362d9

File: e2e/capacity_test.go (hunk 14 above)
Comment (bells17) [reply e2e/capacity_test.go]: Sorry, I fixed it.
I want to check below things in this tests:
- the pod having a PVC that is able to schedule is running by the storage capacity tracking
- the pod having a PVC that is not able to schedule is not running by the storage capacity tracking because the target node doesn't have  enough storage
- when assigned a device class that has enough storage to the PVC, the pod having a PVC that is able to schedule is running by the storage capacity tracking even if the storage capacity of the default device class doesn't have enough storage

**Comment by toshipp** [comment]
@bells17 LGTM.
We rebase and squash before the merge.

**Review by llamerada-jp — APPROVED** [review]
LGTM

**Review by toshipp — APPROVED** [review]
